src/EnergyPlus/Pipes.hh

```diff
@@ -1,129 +1,53 @@
 		int BranchIndex; // Index of plant Branch index where this pipe resides
 		int CompIndex; // Index of plant Comp index where this pipe resides
-		bool OneTimeInit;
-		bool CheckEquipName;
-		bool EnvrnFlag;
 
 		// Default Constructor
 		LocalPipeData() :
-			TypeOf( 0 ),
 			InletNodeNum( 0 ),
 			OutletNodeNum( 0 ),
 			LoopNum( 0 ),
 			LoopSide( 0 ),
 			BranchIndex( 0 ),
-			CompIndex( 0 ),
-			OneTimeInit( true ),
-			CheckEquipName( true ),
-			EnvrnFlag( true )
+			CompIndex( 0 )
 		{}
 
-		// Member Constructor
-		LocalPipeData(
-			std::string const & Name, // main plant (cooling) loop ID
-			int const TypeOf, // type of pipe
-			int const InletNodeNum, // Node number on the inlet side of the plant
-			int const OutletNodeNum, // Node number on the inlet side of the plant
-			int const LoopNum, // Index of plant loop where this pipe resides
-			int const LoopSide, // Index of plant loop side where this pipe resides
-			int const BranchIndex, // Index of plant Branch index where this pipe resides
-			int const CompIndex, // Index of plant Comp index where this pipe resides
-			bool const OneTimeInit,
-			bool const CheckEquipName,
-			bool const EnvrnFlag
-		) :
-			Name( Name ),
-			TypeOf( TypeOf ),
-			InletNodeNum( InletNodeNum ),
-			OutletNodeNum( OutletNodeNum ),
-			LoopNum( LoopNum ),
-			LoopSide( LoopSide ),
-			BranchIndex( BranchIndex ),
-			CompIndex( CompIndex ),
-			OneTimeInit( OneTimeInit ),
-			CheckEquipName( CheckEquipName ),
-			EnvrnFlag( EnvrnFlag )
-		{}
+		static std::shared_ptr<PlantComponent> pipeFactory( std::string objectName );
+		int performEveryTimeInit();
+		int performOneTimeInit();
+		int performBeginEnvrnInit();
+		int simulate();
 
 	};
 
 	// Object Data
-	extern Array1D< LocalPipeData > LocalPipe; // dimension to number of pipes
-
-	// Functions
-
-	void
-	SimPipes(
-		int const CompType,
-		std::string & PipeName,
-		int & CompIndex,
-		Real64 const MaxVolFlowRate,
-		bool const InitLoopEquip,
-		bool const FirstHVACIteration
-	);
-
-	// End Plant Loop Module Driver Subroutines
-	//******************************************************************************
-
-	// Beginning of Plant Loop Module Get Input subroutines
-	//******************************************************************************
-
-	void
-	GetPipeInput();
-
-	// End of Get Input subroutines for the Plant Loop Module
-	//******************************************************************************
-
-	// Beginning Initialization Section of the Plant Loop Module
-	//******************************************************************************
-
-	void
-	InitializePipes(
-		int const PipeType, // Type of Pipe
-		std::string const & PipeName, // Name of Pipe
-		int & PipeNum, // Index into pipe structure for name
-		Real64 const MaxVolFlowRate // unused at present time
-	);
+	extern Array1D< std::shared_ptr< LocalPipeData > > LocalPipe; // dimension to number of pipes
```

Comment:
Wrt containers of smart pointers I will note that I'm finishing up a "collection" framework that is meant to let us work with these sorts of entities more easily. Some of the features of this are:
* Automatic dereferencing of iterators and subscripts to avoid pointer syntax
* Sub-collections are easily generated and lightweight (pointers to elements internally)
* Uniform, simple API
* 1-based indexing (collection(1) is the first element)
* Grow-friendly internal containers (like std::vector) to reduce heap allocations

Sub-collections will be used to improve performance and scalability of performance-critical loops that now loop over all elements but only operate on a subset. This is a big performance hit that gets worse for larger models (big-*O* complexity for you CS types).

Boost pointer containers and indirect iterators provide some of this but not everything I'm going for and bring in a big dependency and (possibly much) slower builds. I'm happy to provide more details.